### DISCUSSION TITLE
Parallel Routes vs iFrames

### DISCUSSION THREAD

**Discussion by laurens-mesure** [vercel/next.js#60681] [Help]
## Context

I'm developing a Web OS at [inuva.me](https://inuva.me/) featuring multiple mini-applications like a file explorer, notepad, and music player. My approach involves server-side rendering of these applications to maintain a lightweight client (bundle).

### Challenges with Parallel routes

I have thought about the most performant way to render multiple self contained mini-applications (file explorer, notepad, music player...) on the "desktop" (located on the "/" route). Parallel routes seem excellent for this but those won't work for my usecase. 

- **Instance Multiplicity:** Need for multiple instances of the same app (e.g., dual file explorers), which doesn't align well with the design of parallel routes.
- **Shared URL Issues:** Parallel routes share a single URL, leading to conflicts in functionalities like search (implemented through searchParams). This causes unintended searches across multiple apps.
- **JSX Recalculation Concern:** Note: I'm not 100% sure of this. Suspecting that a URL change triggers a recalulation of all active parallel routes and the current route, affecting performance.
- **State Management Limitations:** Passing state to parallel routes seems restrictive, only via URL and context providers.

With parallel routes, I had to rely heavily on `revalidatePath("/")`. I know this shouldn't be the case but I could not see a way to implement mutations with a lot of side effects in different mini-applications without `revalidatePath("/")`.

### Using iFrames

I have used iFrames like this: each mini-application (file explorer, notepad, music player) has it's own route under a dynamic route `/[processId]/...`. The root path is the desktop and renders iFrames for each application that is currently running. The iFrame looks like this ``<iFrame path=`/${processId}/explorer` />`` This has resulted in several benefits for my project:

- **Distinct URLs for Apps:** Each app, hosted under dynamic routes like `/[processId]/...`, maintains its own URL and state. This contrasts with the total recalculations required by parallel routes.
- **Deferred Loading Advantage:** Leveraging iFrames allows on-demand loading of applications, reducing initial bundle size.
- **Process Isolation in Chromium:** iFrames run in separate processes, beneficial for my use case despite the complexity in Inter-Process Communication.

#### Drawbacks of iFrames

- **Loss of `revalidatePath` and `revalidateTag` Functions:** iFrames, acting as separate sites, limit the ability to update UI state via these methods.
- **Complexity in Event Bridging:** Event communication between parent and child documents is challenging but feasible with methods like `parent.postMessage`

## Seeking Guidance

- Are there more optimal or intended methods to achieve what I'm doing?
- Is server-side rendering possible for content within an iFrame on the first load?
- Am I approaching this from the wrong direction?

**Comment by TheMikeyRoss** [comment]
iFrames are like Adobe Flash
